Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/polyline_snapping.h

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) 2018 GeometryFactory (France).
+// All rights reserved.
+//
+// This file is part of CGAL (www.cgal.org).
+// You can redistribute it and/or modify it under the terms of the GNU
+// General Public License as published by the Free Software Foundation,
+// either version 3 of the License, or (at your option) any later version.
+//
+// Licensees holding a valid commercial license may use this file in
+// accordance with the commercial license agreement provided with the software.
+//
+// This file is provided AS IS with NO WARRANTY OF ANY KIND, INCLUDING THE
+// WARRANTY OF DESIGN, MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+//
+// $URL$
+// $Id$
+// SPDX-License-Identifier: GPL-3.0+
+//
+//
+// Author(s)     : Simon Giraudot
+
+#ifndef CGAL_POLYGON_MESH_PROCESSING_POLYLINE_SNAPPING_H
+#define CGAL_POLYGON_MESH_PROCESSING_POLYLINE_SNAPPING_H
+
+#include <CGAL/license/Polygon_mesh_processing/corefinement.h>
+
+#include <CGAL/Exact_predicates_exact_constructions_kernel.h>
+#include <CGAL/box_intersection_d.h>
+#include <CGAL/tuple.h>
+
+#include <CGAL/boost/iterator/counting_iterator.hpp>
+#include <boost/range/const_iterator.hpp>
+#include <boost/range/value_type.hpp>
+#include <boost/foreach.hpp>
+
+#include <cmath>
+
+/// \cond SKIP_IN_MANUAL
+
+namespace CGAL {
+namespace Polygon_mesh_processing{
+namespace internal {
+
+template <typename Exact_kernel>
+bool exact_snapping (const typename Exact_kernel::Segment_3& s0,
+                     const typename Exact_kernel::Segment_3& s1,
+                     typename Exact_kernel::Point_3& result)
+{
+  typedef typename Exact_kernel::Point_3 Point_3;
+  typedef typename Exact_kernel::Segment_3 Segment_3;
+  typedef typename Exact_kernel::Vector_3 Vector_3;
+  typedef typename Exact_kernel::Plane_3 Plane_3;
+
+  CGAL_assertion (s0.source() != s0.target());
+  CGAL_assertion (s1.source() != s1.target());
+
+  Vector_3 v0 = s0.to_vector();
+  Vector_3 v1 = s1.to_vector();
+  Vector_3 normal = CGAL::cross_product (v0, v1);
+  if (normal == CGAL::NULL_VECTOR)
+    return false;
+  
+  Plane_3 plane0 (s0.source(), normal);
+  Plane_3 plane1 (s1.source(), normal);
+
+  Segment_3 s0proj (plane1.projection (s0.source()),
+                    plane1.projection (s0.target()));
+  Segment_3 s1proj (plane0.projection (s1.source()),
+                    plane0.projection (s1.target()));
+
+  CGAL_assertion (s0proj.source() != s0proj.target());
+  CGAL_assertion (s1proj.source() != s1proj.target());
+  
+  if (!CGAL::do_intersect (s0, s1proj) ||
+      !CGAL::do_intersect (s1, s0proj))
+    return false;
+
+  Point_3 p0, p1;
+  
+  typename CGAL::cpp11::result_of<typename Exact_kernel::Intersect_3(Segment_3, Segment_3)>::type
+    intersection0 = intersection(s0, s1proj);
+  if (!intersection0)
+    return false;
+
+  if (const Point_3* p = boost::get<Point_3>(&*intersection0))
+    p0 = *p;
+  else
+    return false;
```

Comment:
I don't know what the function is supposed to do but I'm surprized you return false in case the segments are coplanar and collinear